crates/iota-data-ingestion-core/src/reader/fetch.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::BTreeMap,
+    ffi::OsString,
+    fmt::Display,
+    fs,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use iota_rest_api::{CheckpointData, Client};
+use iota_storage::blob::Blob;
+use iota_types::messages_checkpoint::CheckpointSequenceNumber;
+use notify::{RecursiveMode, Watcher};
+use object_store::{ObjectStore, path::Path as ObjectStorePath};
+use tokio::sync::mpsc;
+use tracing::{debug, info};
+
+use crate::{IngestionError, IngestionResult, MAX_CHECKPOINTS_IN_PROGRESS};
+
+pub type CheckpointResult = IngestionResult<(Arc<CheckpointData>, usize)>;
+
+#[derive(Debug, Clone, Copy)]
+pub enum ReadSource {
+    Local,
+    Remote,
+}
+
+impl Display for ReadSource {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ReadSource::Local => write!(f, "local"),
+            ReadSource::Remote => write!(f, "remote"),
+        }
+    }
+}
+
+/// Reads and deserializes unprocessed checkpoint files from the given
+/// directory, up to a capacity limit.
+///
+/// This function scans the specified path for checkpoint files whose sequence
+/// number is greater than or equal to `current_checkpoint_number`. It attempts
+/// to deserialize each file into a [`CheckpointData`], collecting them
+/// into a vector. The process stops early if the provided `exceeds_capacity`
+/// function returns `true` for a checkpoint's sequence number,
+/// or when `MAX_CHECKPOINTS_IN_PROGRESS` files have been processed.
+pub async fn read_local_files_with_capacity_check<F>(
+    path: &Path,
+    current_checkpoint_number: CheckpointSequenceNumber,
+    exceeds_capacity: F,
```

Comment:
Instead of passing a closure as an argument, you could define a trait with a method `exceeds_capacity` that implementors will have to implement and default implementations for the rest top-level functions here.